src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/StateMachineRewriter.cs

```diff
@@ -390,18 +392,33 @@ protected SynthesizedImplementationMethod GenerateIteratorGetEnumerator(MethodSy
             if ((object)initialThreadIdField != null)
             {
                 managedThreadId = MakeCurrentThreadId();
+
+                var thenBuilder = ArrayBuilder<BoundStatement>.GetInstance(4);
+                thenBuilder.Add(
+                    // this.state = {initialState};
+                    F.Assignment(F.Field(F.This(), stateField), F.Literal(initialState)));
+
+                thenBuilder.Add(
+                    // result = this;
+                    F.Assignment(F.Local(resultVariable), F.This()));
+
+                var extraReset = GetExtraResetForIteratorGetEnumerator();
+                if (extraReset != null)
```

Comment:
>extraReset != null [](start = 20, length = 18)

Would it make sense to pass in the builder to `GetExtraResetForIteratorGetEnumerator()`? (That would also allow adding multiple statements, although we probably don't have such a scenario today.)

Reply:
That can be refactored easily when the need arises.

---
In reply to: [245383654](https://github.com/dotnet/roslyn/pull/31764#discussion_r245383654) [](ancestors = 245383654)